### PR TITLE
chore(cloudformation): add ParseTemplateDescriptions func

### PR DESCRIPTION
**Pull request by efekarakus**
`ParseTemplateDescriptions` parses a CloudFormation template and
returns the comments associated with a resource. We'll use these
comments to display human-readable text during deployments.

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: internal/pkg/aws/cloudformation/testdata/parse/env.yaml [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+          Value: !Sub 'copilot-${AppName}-${EnvironmentName}'
+
+  InternetGatewayAttachment:
+    Type: AWS::EC2::VPCGatewayAttachment
+    Properties:
+      InternetGatewayId: !Ref InternetGateway
+      VpcId: !Ref VPC
+
+  PublicSubnet1:
+  # A public subnet in your first AZ for internet facing services.
```
Comment (huanjani) [inline internal/pkg/aws/cloudformation/testdata/parse/env.yaml:42]: Do we need to write out "availability zone (AZ)"? Or is that abbreviation commonly known? (I assume the real yamls will be similar to the testdata yamls.)

Comment (efekarakus) [reply internal/pkg/aws/cloudformation/testdata/parse/env.yaml:42]: This file is only for testing (anything under ["testdata"](https://dave.cheney.net/2016/05/10/test-fixtures-in-go)) so I've put whatever I could think of at the time.

I'll send a separate PR adding officially these comments to our templates after this

**Review by huanjani — APPROVED** [review]
Neat!

File: internal/pkg/aws/cloudformation/parse.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	descriptionFor = make(map[string]string)
+	for i := 0; i < len(tpl.Resources.Content); i += 2 {
+		// The content of a map, like the "Resources" field, always come in pairs.
+		// The first element represents a key, ex: {Value: "Cluster", Kind: ScalarNode, Tag: "!!str", Content: nil}
+		// The second element holds the properties, ex: {Value: "", Kind: MappingNode, Tag:"!!map", Content:[...]}
+		resource := tpl.Resources.Content[i]
+		properties := tpl.Resources.Content[i+1]
+		description := properties.Content[0].HeadComment // The description is the comment above the first property.
```
Comment (iamhopaul123) [inline internal/pkg/aws/cloudformation/parse.go]: Is it possible to get out of index exception?

File: internal/pkg/aws/cloudformation/parse.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		if description == "" {
+			continue
+		}
+
+		logicalID := resource.Value
+		descriptionFor[logicalID] = strings.Trim(description, " #" /* remove both spaces and the # character */)
```
Comment (iamhopaul123) [inline internal/pkg/aws/cloudformation/parse.go]: Is it possible to have more than one space before #?

File: internal/pkg/aws/cloudformation/parse.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	var tpl template
+	if err := yaml.Unmarshal([]byte(body), &tpl); err != nil {
+		return nil, fmt.Errorf("unmarshal cloudformation template: %w", err)
+	}
+
+	descriptionFor = make(map[string]string)
+	for i := 0; i < len(tpl.Resources.Content); i += 2 {
+		// The content of a map, like the "Resources" field, always come in pairs.
+		// The first element represents a key, ex: {Value: "Cluster", Kind: ScalarNode, Tag: "!!str", Content: nil}
+		// The second element holds the properties, ex: {Value: "", Kind: MappingNode, Tag:"!!map", Content:[...]}
```
Comment (iamhopaul123) [inline internal/pkg/aws/cloudformation/parse.go]: Black magic again 🤔

**Review by iamhopaul123 — COMMENTED** [review]
This is so cool! However, can we try to make sure our unit tests cover more edge cases so that we can do best parsing and handle the exception properly when the format is not what we typically want.

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/aws/cloudformation/parse.go (hunk 2 above)
Comment (efekarakus) [reply internal/pkg/aws/cloudformation/parse.go]: A [resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resources-section-structure.html) in CFN always has a `Type` and `Properties` field so it shouldn't happen unless the client does not pass a cfn template.
~~But in that situation the unmarshal call should fail.~~ But in that situation `Resources` will be empty so we won't go through this loop.

Comment (efekarakus) [reply internal/pkg/aws/cloudformation/parse.go]: I renamed the variables like this, I think it makes it a bit more explicit
```go
		logicalID := tpl.Resources.Content[i]
		fields := tpl.Resources.Content[i+1]
```

File: internal/pkg/aws/cloudformation/parse.go (hunk 3 above)
Comment (efekarakus) [reply internal/pkg/aws/cloudformation/parse.go]: Sorry this comment didn't clarify things at all. I reworded it like this:
```suggestion
		descriptionFor[logicalID] = strings.Trim(description, " #" /* remove leading and trailing chars {" ", "#"} */)
```
The function signature for `Trim` is really strange: https://golang.org/pkg/strings/#Trim. The second parameter is a **set** that's represented as a string, so this is saying remove both all leading and tailing spaces and hash signs from the string.
